src/sync/chase_lev.rs

```diff
@@ -254,6 +273,40 @@ impl<T> Deque<T> {
         }
     }
 
+    fn steal_half(&self) -> StealHalf<T> {
+        let guard = epoch::pin();
+
+        let t = self.top.load(Acquire);
+        fence(SeqCst);
+        let b = self.bottom.load(Acquire);
+
+        let size = b - t;
+        if size <= 0 {
+            return StealHalf::Empty;
+        }
+
+        let half = (size + 1) / 2;
+
+        unsafe {
+            let a = self.array.load(Acquire, &guard).unwrap();
+
+            let mut data = Vec::with_capacity(half as usize);
+            for i in t..t + half {
+                data.push(a.get(i));
+            }
+
+            if self.top.compare_and_swap(t, t + half, SeqCst) == t {
```

Comment:
What happens if just before this CAS the worker thread pops most of the elements, so that `t + half` overshoots the bottom?

There is a special if-statement in `fn pop` where we deal with a special racy case when popping the last element. With the presence of the steal-half operation, now `steal_half` and `pop` can race in more complex ways.

The racy case in `fn pop` is currently handled by incrementing top rather than decrementing bottom. With steal-half we can't do a similar trick.

Perhaps this is the difficulty with implementing steal-half we've been missing all along? :)